File: rust/crates/core/src/voucher.rs
//! Wire-agnostic payment-channel voucher verification.
//!
//! A voucher authorizes a cumulative spend on a channel: the 48-byte Borsh
//! payload `channelId ‖ cumulativeAmount ‖ expiresAt` (see
//! [`crate::payment_channels::voucher_message_bytes`]) signed Ed25519 by the
//! channel's authorized signer. Shared by the MPP `session` intent and the x402
//! `batch-settlement` scheme.

use std::str::FromStr;

use solana_pubkey::Pubkey;

use crate::payment_channels::voucher_message_bytes;
use crate::{Error, Result};

/// Verify an Ed25519 voucher signature against the authorized signer and check
/// that its expiry is acceptable.
///
/// - `channel_id`, `signature_b58`, `authorized_signer_b58` are base58.
/// - `now` is the current Unix time in seconds.
/// - `settlement_window` is the channel's forced-close grace period (seconds).
///   A non-zero `expires_at` MUST outlast the settlement window — i.e. it must
///   be at least `now + settlement_window` — so the operator can still redeem
///   the voucher on-chain after the asynchronous settlement delay. Pass `0` to
///   only require the voucher be unexpired (`expires_at > now`).
///
/// Expiry semantics mirror the on-chain program (`expires_at != 0 && now >=
/// expires_at` rejects): **`expires_at == 0` means the voucher never expires**
/// and is always accepted regardless of `now` or `settlement_window`. A
/// non-zero `expires_at` is rejected when it is at/before `now` (expired) or
/// before `now + settlement_window` (would lapse before settlement).
pub fn verify_voucher_signature(
    channel_id: &str,
    cumulative: u64,
    expires_at: i64,
    signature_b58: &str,
    authorized_signer_b58: &str,
    now: i64,
    settlement_window: i64,
) -> Result<()> {
    use ed25519_dalek::{Signature, VerifyingKey};

    // `expires_at == 0` is never-expires (on-chain parity); skip all time checks.
    if expires_at != 0 {
        if expires_at <= now {
            return Err(Error::Other("Voucher has expired".to_string()));
        }
        // Must outlast the forced-close grace period so it can still settle
        // on-chain after the async settlement delay. `now + settlement_window`
        // is computed with saturating math to avoid i64 overflow.
        let settle_deadline = now.saturating_add(settlement_window.max(0));
        if expires_at < settle_deadline {
            return Err(Error::Other(format!(
                "Voucher expiry {expires_at} does not outlast the settlement window \
                 (must be >= {settle_deadline} = now {now} + window {settlement_window})"
            )));
        }
    }

    let channel = Pubkey::from_str(channel_id)
        .map_err(|e| Error::Other(format!("invalid channelId: {e}")))?;
    let message = voucher_message_bytes(&channel, cumulative, expires_at)?;

    let sig_bytes = bs58::decode(signature_b58)
        .into_vec()
        .map_err(|e| Error::Other(format!("Invalid signature encoding: {e}")))?;
    let pubkey_bytes = bs58::decode(authorized_signer_b58)
        .into_vec()
        .map_err(|e| Error::Other(format!("Invalid authorized_signer: {e}")))?;

    let key_arr: [u8; 32] = pubkey_bytes
        .try_into()
        .map_err(|_| Error::Other("Pubkey is not 32 bytes".to_string()))?;
    let sig_arr: [u8; 64] = sig_bytes
        .try_into()
        .map_err(|_| Error::Other("Signature is not 64 bytes".to_string()))?;

    let verifying_key = VerifyingKey::from_bytes(&key_arr)
        .map_err(|e| Error::Other(format!("Invalid authorized_signer key: {e}")))?;
    let signature = Signature::from_bytes(&sig_arr);

    verifying_key
        .verify_strict(&message, &signature)
        .map_err(|_| Error::Other("Voucher signature verification failed".to_string()))
}

#[cfg(test)]
mod tests {
    use super::*;
    use ed25519_dalek::{Signer, SigningKey};

    fn signed(
        channel: &Pubkey,
        cumulative: u64,
        expires_at: i64,
        sk: &SigningKey,
    ) -> (String, String) {
        let msg = voucher_message_bytes(channel, cumulative, expires_at).unwrap();
        let sig = sk.sign(&msg);
        let signer = bs58::encode(sk.verifying_key().as_bytes()).into_string();
        (bs58::encode(sig.to_bytes()).into_string(), signer)
    }

    #[test]
    fn accepts_a_valid_unexpired_voucher_and_rejects_tampering() {
        let sk = SigningKey::from_bytes(&[7u8; 32]);
        let channel = Pubkey::new_unique();
        let channel_b58 = bs58::encode(channel.as_ref()).into_string();
        let (sig, signer) = signed(&channel, 1_000, 4_102_444_800, &sk);

        assert!(
            verify_voucher_signature(&channel_b58, 1_000, 4_102_444_800, &sig, &signer, 1, 0)
                .is_ok()
        );
        // Wrong cumulative → signature no longer matches the message.
        assert!(
            verify_voucher_signature(&channel_b58, 2_000, 4_102_444_800, &sig, &signer, 1, 0)
                .is_err()
        );
        // Expired.
        assert!(verify_voucher_signature(&channel_b58, 1_000, 100, &sig, &signer, 200, 0).is_err());
    }

    // FIX #9a: expires_at == 0 means never-expires (on-chain parity), so it is
    // accepted regardless of `now` or the settlement window.
    #[test]
    fn zero_expiry_means_never_expires() {
        let sk = SigningKey::from_bytes(&[8u8; 32]);
        let channel = Pubkey::new_unique();
        let channel_b58 = bs58::encode(channel.as_ref()).into_string();
        let (sig, signer) = signed(&channel, 1_000, 0, &sk);

        // Accepted even with a large `now` and a large settlement window.
        assert!(verify_voucher_signature(
            &channel_b58,
            1_000,
            0,
            &sig,
            &signer,
            1_000_000_000,
            900
        )
        .is_ok());
    }

    // FIX #9b: a non-zero expiry must outlast the settlement window.
    #[test]
    fn nonzero_expiry_must_outlast_settlement_window() {
        let sk = SigningKey::from_bytes(&[9u8; 32]);
        let channel = Pubkey::new_unique();
        let channel_b58 = bs58::encode(channel.as_ref()).into_string();
        let now = 1_000;
        let window = 900;

        // expires_at just inside the window → rejected (would lapse before settle).
        let (sig_short, signer) = signed(&channel, 1_000, now + window - 1, &sk);
        let err = verify_voucher_signature(
            &channel_b58,
            1_000,
            now + window - 1,
            &sig_short,
            &signer,
            now,
            window,
        )
        .unwrap_err();
        assert!(
            err.to_string().contains("settlement window"),
            "expected settlement-window error, got: {err}"
        );

        // expires_at exactly at now + window → accepted.
        let (sig_ok, _) = signed(&channel, 1_000, now + window, &sk);
        assert!(verify_voucher_signature(
            &channel_b58,
            1_000,
            now + window,
            &sig_ok,
            &signer,
            now,
            window,
        )
        .is_ok());

        // expired (at/before now) → rejected even with a zero window.
        let (sig_exp, _) = signed(&channel, 1_000, now, &sk);
        assert!(
            verify_voucher_signature(&channel_b58, 1_000, now, &sig_exp, &signer, now, 0).is_err()
        );
    }
}

File: typescript/packages/mpp/src/server/session/voucher.ts
// Voucher verifier for the MPP session server.
//
// Pure function — given a current channel snapshot and a signed voucher,
// decide whether to accept (and what the new watermark would be), reject,
// or treat as an idempotent replay. The caller persists any accepted
// delta through `store.updateChannel` (typically inside its own
// re-read-and-recheck closure to be safe under concurrency).
//
// Mirrors the 9-step verifier in
// `rust/crates/mpp/src/server/session.rs::SessionServer::verify_voucher`
// (lines ~427–543). Step numbering below matches the Rust source.

import type { SignedVoucher } from '../../shared/session-types.js';
import { verifyVoucherSignature } from '../../shared/voucher.js';
import type { ChannelState } from './store.js';

/**
 * Reasons a voucher can be rejected. Stable string tags so the caller
 * can map to HTTP statuses / log levels without parsing free text.
 */
export type VoucherRejectReason =
    | 'below-min-delta'
    | 'channel-close-pending'
    | 'channel-sealed'
    | 'cumulative-not-monotonic'
    | 'exceeds-deposit'
    | 'expired'
    | 'expires-within-settlement-window'
    | 'invalid-cumulative'
    | 'invalid-signature';

/** Verification outcome: the voucher advanced the channel watermark. */
export interface VoucherVerifyAccepted {
    /** New cumulative watermark to persist. */
    readonly newCumulative: bigint;
    /** Expiry of the now-highest voucher (i64 seconds). */
    readonly newExpiresAt: bigint;
    /** Signature to persist as `highestVoucherSignature`. */
    readonly newSignature: string;
    readonly status: 'accepted';
}

/** Verification outcome: an already-accepted voucher was re-submitted (idempotent). */
export interface VoucherVerifyReplayed {
    /** Existing watermark, returned for caller convenience. */
    readonly newCumulative: bigint;
    readonly status: 'replayed';
}

/** Verification outcome: the voucher was rejected. See {@link VoucherRejectReason}. */
export interface VoucherVerifyRejected {
    /** Human-readable detail. Safe to log; not stable. */
    readonly detail: string;
    readonly reason: VoucherRejectReason;
    readonly status: 'rejected';
}

/** Union of the three voucher verification outcomes. */
export type VoucherVerifyResult = VoucherVerifyAccepted | VoucherVerifyRejected | VoucherVerifyReplayed;

/** Arguments to {@link verifyVoucherForChannel}. */
export interface VerifyVoucherArgs {
    /**
     * Authoritative deposit cap. Passed in (rather than read off `state`)
     * because some callers carry an updated cap after a recent top-up
     * that hasn't yet been written back into the store.
     */
    readonly deposit: bigint;
    /** Optional minimum delta from the previous cumulative. */
    readonly minVoucherDelta?: bigint | undefined;
    /**
     * Unix seconds, defaults to `Math.floor(Date.now() / 1000)`. Exposed
     * for tests and for callers that already have a clock.
     */
    readonly nowSeconds?: bigint | undefined;
    /**
     * Forced-close grace period of the channel, in seconds (the on-chain
     * `gracePeriod` captured at open). When set, a *non-zero* voucher
     * `expiresAt` must outlast the settlement window — i.e. it must still
     * be valid `settlementWindow` seconds from now, so the merchant can
     * land an async settle_and_seal before the voucher expires.
     * A voucher with `expiresAt == 0` never expires and is unaffected.
     * Defaults to 0 (window check disabled — only `expiresAt <= now` is
     * rejected for non-zero expiries).
     */
    readonly settlementWindow?: bigint | undefined;
    /** Voucher being submitted. */
    readonly signed: SignedVoucher;
    /** Channel snapshot — typically read just before calling. */
    readonly state: ChannelState;
}

/**
 * Verify a voucher against a channel snapshot.
 *
 * Returns a verdict; the caller is responsible for persisting any
 * accepted delta via `store.updateChannel`. The verifier is pure —
 * no store, network, or clock side effects (clock is injectable).
 *
 * Mirrors `SessionServer::verify_voucher` in Rust. Expiry is checked
 * here rather than inside the signature helper (Rust intermixes the
 * two; we keep them separate so callers can override `nowSeconds`).
 */
export async function verifyVoucherForChannel(args: VerifyVoucherArgs): Promise<VoucherVerifyResult> {
    const { state, signed, deposit } = args;
    const { data } = signed;

    // 1. Parse new_cumulative from payload
    let newCumulative: bigint;
    try {
        newCumulative = parseU64(data.cumulativeAmount);
    } catch (error) {
        return reject('invalid-cumulative', errorMessage(error));
    }

    // 2. Channel must not be sealed
    if (state.sealed) {
        return reject('channel-sealed', `Channel ${state.channelId} is already sealed`);
    }

    // 3. Channel must not be in close-pending
    if (state.closeRequestedAt !== undefined) {
        return reject(
            'channel-close-pending',
            `Channel ${state.channelId} close is pending — no further vouchers accepted`,
        );
    }

    // 4. Idempotent replay: same cumulative AND same signature
    if (newCumulative === state.cumulative && state.highestVoucherSignature === signed.signature) {
        // Rust still re-verifies the signature on the replay path. We do
        // the same so a replay of a forged voucher can't slip through.
        const ok = await safeVerifySignature(signed, state.authorizedSigner);
        if (!ok.ok) return ok.reject;
        const expiryReject = checkExpiry(toBigInt(data.expiresAt), currentTime(args.nowSeconds), args.settlementWindow);
        if (expiryReject) return expiryReject;
        return { newCumulative, status: 'replayed' };
    }

    // 5. Must strictly exceed watermark (non-replay case)
    if (newCumulative <= state.cumulative) {
        return reject(
            'cumulative-not-monotonic',
            `Voucher cumulative ${newCumulative} must exceed watermark ${state.cumulative}`,
        );
    }

    // 6. Must not exceed deposit
    if (newCumulative > deposit) {
        return reject('exceeds-deposit', `Voucher cumulative ${newCumulative} exceeds deposit ${deposit}`);
    }

    // 7. Min delta check
    const delta = newCumulative - state.cumulative;
    const minDelta = args.minVoucherDelta ?? 0n;
    if (minDelta > 0n && delta < minDelta) {
        return reject('below-min-delta', `Voucher delta ${delta} is below minimum ${minDelta}`);
    }

    // 8. Verify signature (Ed25519 over the 50-byte canonical payload)
    const sigCheck = await safeVerifySignature(signed, state.authorizedSigner);
    if (!sigCheck.ok) return sigCheck.reject;

    // 9. Expiry — caller may override `nowSeconds` for deterministic tests.
    const expiresAt = toBigInt(data.expiresAt);
    const expiryReject = checkExpiry(expiresAt, currentTime(args.nowSeconds), args.settlementWindow);
    if (expiryReject) return expiryReject;

    return {
        newCumulative,
        newExpiresAt: expiresAt,
        newSignature: signed.signature,
        status: 'accepted',
    };
}

// ── helpers ──

function reject(reason: VoucherRejectReason, detail: string): VoucherVerifyRejected {
    return { detail, reason, status: 'rejected' };
}

/**
 * Voucher expiry rule, mirroring the on-chain settle check
 * (`expires_at != 0 && now >= expires_at` ⇒ rejected) so the off-chain
 * acceptance never advances a watermark the program would later refuse.
 *
 * - `expiresAt == 0` ⇒ never expires; always accepted regardless of `now`.
 * - non-zero `expiresAt <= now` ⇒ already expired.
 * - non-zero `expiresAt` within `settlementWindow` of `now` ⇒ rejected: it
 *   would not outlast the forced-close grace period, so a merchant settle
 *   could miss the window. Skipped when `settlementWindow` is unset/zero.
 *
 * Returns a {@link VoucherVerifyRejected} to short-circuit on, or
 * `undefined` when the voucher's expiry is acceptable.
 */
function checkExpiry(
    expiresAt: bigint,
    now: bigint,
    settlementWindow: bigint | undefined,
): VoucherVerifyRejected | undefined {
    // 0 means never-expires on-chain — accept unconditionally.
    if (expiresAt === 0n) return undefined;
    if (expiresAt <= now) {
        return reject('expired', `Voucher expired at ${expiresAt} (now ${now})`);
    }
    const window = settlementWindow ?? 0n;
    if (window > 0n && expiresAt < now + window) {
        return reject(
            'expires-within-settlement-window',
            `Voucher expiresAt ${expiresAt} is within the settlement window (now ${now} + window ${window} = ${now + window}); ` +
                'it may expire before settlement lands',
        );
    }
    return undefined;
}

async function safeVerifySignature(
    signed: SignedVoucher,
    authorizedSigner: string,
): Promise<{ ok: false; reject: VoucherVerifyRejected } | { ok: true }> {
    try {
        const valid = await verifyVoucherSignature({
            signatureBase58: signed.signature,
            signerBase58: authorizedSigner,
            voucher: signed.data,
        });
        if (!valid) {
            return { ok: false, reject: reject('invalid-signature', 'Voucher signature verification failed') };
        }
        return { ok: true };
    } catch (error) {
        return {
            ok: false,
            reject: reject('invalid-signature', errorMessage(error)),
        };
    }
}

function parseU64(value: string): bigint {
    if (!/^\d+$/.test(value)) {
        throw new Error(`Invalid cumulative in voucher: ${value}`);
    }
    const parsed = BigInt(value);
    if (parsed < 0n || parsed > (1n << 64n) - 1n) {
        throw new Error(`Cumulative ${value} outside u64 range`);
    }
    return parsed;
}

function toBigInt(value: bigint | number | string): bigint {
    if (typeof value === 'bigint') return value;
    if (typeof value === 'number') {
        if (!Number.isSafeInteger(value)) throw new Error(`expiresAt is not a safe integer: ${value}`);
        return BigInt(value);
    }
    return BigInt(value);
}

function currentTime(override: bigint | undefined): bigint {
    if (override !== undefined) return override;
    return BigInt(Math.floor(Date.now() / 1000));
}

function errorMessage(error: unknown): string {
    if (error instanceof Error) return error.message;
    return String(error);
}
